### PR TITLE
Update Onebranch pipelines to reference templates

### PR DESCRIPTION
**Pull request by mikeandescavage**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9772377

### What this PR does / why we need it:

These changes will reduce the number of pipeline definitions that will need to be updated by referencing templates. A change to the template will then be applied to all three.  However, if any jobs/stages need to be added to the base pipeline, those changes will also have to be made manually on the remaining pull/official pipeline definitions.

### Test plan for issue:

Tested both development pipelines against the templates
https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=41724395&view=results
https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=41724291&view=results

### Is there any documentation that needs to be updated for this PR?

Clean up / tech debt

### CODE REVIEW

File: .pipelines/onebranch/pipeline.buildrp.dev.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# Support:        https://aka.ms/onebranchsup                                   #
+#################################################################################
+
+trigger: none
+pr: none
+
+variables:
+  CDP_DEFINITION_BUILD_COUNT: $[counter('', 0)] # needed for onebranch.pipeline.version task https://aka.ms/obpipelines/versioning
+  LinuxContainerImage: cdpxlinux.azurecr.io/global/ubuntu-1804-all:5.0   # Docker image which is used to build the project https://aka.ms/obpipelines/containers
+  DEBIAN_FRONTEND: noninteractive
```
Comment (mjudeikis) [inline .pipelines/onebranch/pipeline.buildrp.dev.yml:17]: Can we use one variable format? We use CamelCase around the place

Comment (mjudeikis) [reply .pipelines/onebranch/pipeline.buildrp.dev.yml:17]: And these are not used. I assume this is somehow related to 
```
extends:
  template: v2/OneBranch.NonOfficial.CrossPlat.yml@templates # https://aka.ms/obpipelines/templates
```
Would be good to understand how.

File: .pipelines/onebranch/pipeline.image.mirror.dev.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+- name: vsoDeployerBuildID
+  type: string
+  default: latest
+
+variables:
+  CDP_DEFINITION_BUILD_COUNT: $[counter('', 0)] # needed for onebranch.pipeline.version task https://aka.ms/obpipelines/versioning
+  LinuxContainerImage: centos:centos7   # Docker image which is used to build the project https://aka.ms/obpipelines/containers
+  DEBIAN_FRONTEND: noninteractive
+  vso-deployer-pipeline-id: 206753
+  vso-project-id: 5d69ab04-7ded-49dc-84d5-bbbcac4add8d
```
Comment (mjudeikis) [inline .pipelines/onebranch/pipeline.image.mirror.dev.yml:24]: Is these right? Looks hardcoded values

File: .pipelines/onebranch/pipeline.image.mirror.dev.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+pr: none
+
+parameters:
+- name: vsoDeployerBuildID
+  type: string
+  default: latest
+
+variables:
+  CDP_DEFINITION_BUILD_COUNT: $[counter('', 0)] # needed for onebranch.pipeline.version task https://aka.ms/obpipelines/versioning
+  LinuxContainerImage: centos:centos7   # Docker image which is used to build the project https://aka.ms/obpipelines/containers
```
Comment (mjudeikis) [inline .pipelines/onebranch/pipeline.image.mirror.dev.yml:21]: I seen ubuntu, debian, now centos. All for RHEL based product. 
@jewzaam this does not look right or sustainable for a long run

File: .pipelines/onebranch/templates/template-buildrp-buildaro.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    GOPATH: $(Agent.TempDirectory)
+- task: Bash@3
+  displayName: ⚙️ Make ARO
+  inputs:
+    targetType: inline
+    script: |
+      mkdir -p $(Agent.TempDirectory)/src/github.com/Azure/
+      cp -rd $(Build.SourcesDirectory) $(Agent.TempDirectory)/src/github.com/Azure/ARO-RP
+      cd $(Agent.TempDirectory)/src/github.com/Azure/ARO-RP
+      ls
```
Comment (mjudeikis) [inline .pipelines/onebranch/templates/template-buildrp-buildaro.yml:15]: remove?

File: .pipelines/onebranch/templates/template-buildrp-buildaro.yml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      git rev-parse --short HEAD
+      git status --porcelain
```
Comment (mjudeikis) [inline .pipelines/onebranch/templates/template-buildrp-buildaro.yml:17]: remove?

File: .pipelines/onebranch/templates/template-buildrp-builddocker.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+steps:
+- task: onebranch.pipeline.imagebuildinfo@1
+  displayName: Build Multi Stage Dockerfile
+  inputs:
+    repositoryName: aro-rp
+    dockerFileRelPath: ./Dockerfile.aro-multistage
+    dockerFileContextPath: ./
+    registry: cdpxlinux.azurecr.io
```
Comment (mjudeikis) [inline .pipelines/onebranch/templates/template-buildrp-builddocker.yml:8]: Handcoded?

File: .pipelines/onebranch/pipeline.image.mirror.dev.yml (hunk 3 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/pipeline.image.mirror.dev.yml:21]: There are a few things at play here and we can have a longer conversation on the side, but in short we have to work within the Onebranch ecosystem.  We left the mirroring pipeline at the default linux container of centos as it's not generating any binaries, just combining binaries to produce artifacts and manifests for Ev2 to consume.

File: .pipelines/onebranch/templates/template-buildrp-buildaro.yml (hunk 5 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/templates/template-buildrp-buildaro.yml]: Thanks - left over from previous debugging efforts to determine why binaries were being tagged "-dirty".  These have been removed.

File: .pipelines/onebranch/pipeline.image.mirror.dev.yml (hunk 2 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/pipeline.image.mirror.dev.yml:24]: See https://github.com/Azure/ARO-RP/pull/1451#discussion_r617800236

File: .pipelines/onebranch/pipeline.buildrp.dev.yml (hunk 1 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/pipeline.buildrp.dev.yml:17]: I am not against consistency - these were part of the Onebranch pipeline template and were left untouched.  I can test to see if it breaks anything downstream.

File: .pipelines/onebranch/templates/template-buildrp-builddocker.yml (hunk 6 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/templates/template-buildrp-builddocker.yml:8]: Sorry, what is the question here?  Is the concern that `Dockerfile.aro-multistage` is hardcoded here?

File: .pipelines/onebranch/pipeline.buildrp.dev.yml (hunk 1 above)
Comment (mjudeikis) [reply .pipelines/onebranch/pipeline.buildrp.dev.yml:17]: From what im reading it should work both ways. All variables can be used either in ENV format or `$(foo)`, depending on how they accesses. But would not be first time Azure devops docs are misleading. 

https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#environment-variables

File: .pipelines/onebranch/pipeline.image.mirror.dev.yml (hunk 2 above)
Comment (mjudeikis) [reply .pipelines/onebranch/pipeline.image.mirror.dev.yml:24]: :facepalm:

File: .pipelines/onebranch/templates/template-buildrp-builddocker.yml (hunk 6 above)
Comment (mjudeikis) [reply .pipelines/onebranch/templates/template-buildrp-builddocker.yml:8]: Registry name, should this not be variable, or parameter?

**Comment by mjudeikis** [comment]
@jewzaam This looks very domain specific. Ok to merge but need to make sure we have all the knowhow in the team as we will get into same issue as with our old pipelines where it so complicated and without knowhow we had to refactor.

File: .pipelines/onebranch/templates/template-buildrp-builddocker.yml (hunk 6 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/templates/template-buildrp-builddocker.yml:8]: The registry here doesn't matter for our purposes, though we still need to provide a value and of which we left it at the default value.  We update the tag and registry on the Ev2 side depending on the environment we're deploying to (int, prod, etc.)

The workflow, in short, is to build the docker image and publish it as a build artifact.  Ev2 will then consume, retag, and push to the appropriate registries.

https://onebranch.visualstudio.com/OneBranch/_wiki/wikis/OneBranch.wiki/4601/Build-Docker-Images

File: .pipelines/onebranch/pipeline.image.mirror.dev.yml (hunk 2 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/pipeline.image.mirror.dev.yml]: Moved these vars into the template to reduce the number of instances where its hardcoded.  We can revisit with making them edit-able next version.

File: .pipelines/onebranch/pipeline.buildrp.dev.yml (hunk 1 above)
Comment (mikeandescavage) [reply .pipelines/onebranch/pipeline.buildrp.dev.yml]: The big question was whether or not the Onebranch containers cared.  Tested and it works fine after changing to camel case.

**Comment by jewzaam** [comment]
@mjudeikis we have a meeting getting setup to share knowledge in this space.  If you're OK with merging could you take care of that in your day tomorrow?  I wanted to look at this today but have run out of time.  Thank you!

**Review by mjudeikis — APPROVED** [review]

**Comment by mjudeikis** [comment]
@jewzaam sorry gh notifiations got filled :D

**Pull request merged**